### PR TITLE
[A2-803] Added UpdateRule API

### PR DESCRIPTION
**Pull request by tylercloke**
### :nut_and_bolt: Description

Added UpdateRule API

### :athletic_shoe: Demo Script / Repro Steps

```
# curl -kH "api-token: $TOK" https://localhost/apis/iam/v2beta/projects -d "$(jq -n '{ id: "foo-project", name: "my foo project" }')"

# create a rule to modify

# curl -kH "api-token: $TOK" https://localhost/apis/iam/v2beta/rules -d "$(jq -n '{ id: "foo-rule", name: "my foo rule", type: "NODE", project_id: "foo-project", conditions: [{ type: "CHEF_SERVERS", operator: "MEMBER_OF", values: ["prod", "staging"]}]}')" | jq .

# curl -kH "api-token: $TOK" -XPUT https://localhost/apis/iam/v2beta/rules/foo-rule -d "$(jq -n '{ name: "updated name", type: "NODE", project_id: "foo-project", conditions: [{ type: "CHEF_TAGS", operator: "EQUALS", values: ["new tag"]}, { type: "CHEF_SERVERS", operator: "MEMBER_OF", values: ["other stuff", "new condition servers"]}]}')" | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   628  100   262  100   366   7705  10764 --:--:-- --:--:-- --:--:-- 18470
{
  "rule": {
    "id": "foo-rule",
    "project_id": "foo-project",
    "name": "updated name",
    "type": "NODE",
    "conditions": [
      {
        "type": "CHEF_TAGS",
        "values": [
          "new tag"
        ],
        "operator": "EQUALS"
      },
      {
        "type": "CHEF_SERVERS",
        "values": [
          "other stuff",
          "new condition servers"
        ],
        "operator": "MEMBER_OF"
      }
    ]
  }
}

# create some other project and then attempt but fail with an error to change the rule's project

# curl -kH "api-token: $TOK" https://localhost/apis/iam/v2beta/projects -d "$(jq -n '{ id: "foo-project-2", name: "my foo project" }')"

# curl -kH "api-token: $TOK" -XPUT https://localhost/apis/iam/v2beta/rules/foo-rule -d "$(jq -n '{ name: "updated name", type: "NODE", project_id: "foo-project-2", conditions: [{ type: "CHEF_TAGS", operator: "EQUALS", values: ["new tag"]}, { type: "CHEF_SERVERS", operator: "MEMBER_OF", values: ["other stuff", "new condition servers"]}]}')" | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   543    0   175  100   368   9722  20444 --:--:-- --:--:-- --:--:-- 30166
{
  "error": "cannot change project_id for existing rule with ID \"foo-rule\" ",
  "message": "cannot change project_id for existing rule with ID \"foo-rule\" ",
  "code": 9,
  "details": []
}
```

### :chains: Related Resources

Rebase after this merges: https://github.com/chef/automate/pull/458

### :white_check_mark: Checklist

- [ ] Necessary tests added/updated?
- [ ] Necessary docs added/updated?
- [ ] Code actually executed?
- [ ] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/automate-gateway/handler/iam/v2beta/rules/rules.go [hunk 1]
```diff
@@ -90,6 +108,22 @@ func fromExternal(req *pb_req.CreateRuleReq) (*authz.CreateRuleReq, error) {
 	}, nil
 }
 
+// Wish we had generics
+func fromExternalUpdate(req *pb_req.UpdateRuleReq) (*authz.UpdateRuleReq, error) {
+	t, err := fromExternalType(req.Type)
+	if err != nil {
+		return nil, err
+	}
+	conditions, err := fromExternalConditions(req.Conditions)
```
Comment (srenatus) [inline components/automate-gateway/handler/iam/v2beta/rules/rules.go]: So, what about that `err`? 😉

File: components/automate-gateway/handler/iam/v2beta/rules/rules.go [hunk 2]
```diff
@@ -90,6 +108,22 @@ func fromExternal(req *pb_req.CreateRuleReq) (*authz.CreateRuleReq, error) {
 	}, nil
 }
 
+// Wish we had generics
```
Comment (srenatus) [inline components/automate-gateway/handler/iam/v2beta/rules/rules.go]: 💯 the 1970s called, they want their language back

File: components/authz-service/server/v2/rules_test.go [hunk 3]
```diff
@@ -157,6 +157,149 @@ func TestCreateRule(t *testing.T) {
...
+			resp, err := cl.UpdateRule(ctx, &api.UpdateRuleReq{
+				Id:         "not-found",
+				Name:       "my other foo",
+				ProjectId:  "bar",
+				Conditions: apiConditions,
+			})
+			grpctest.AssertCode(t, codes.NotFound, err)
+			assert.Nil(t, resp)
+		}},
+		{"if the passed rule changes the project, returns 'failed precondition'", func(t *testing.T) {
```
Comment (bcmdarroch) [inline components/authz-service/server/v2/rules_test.go]: you can't change the rule type, but what if you change your conditions so they no longer match the correct type?
```
NODE ATTRIBUTES
Chef Organization (string)
Chef Server (string)
Environment (string)
Chef Role (string[])
Chef Tag (string[])
Chef Policy Name (string)
Chef Policy Group (string)

EVENT ATTRIBUTES
Chef Organization (string)
Chef Server (string)
```

so if I'm editing an Event rule, I shouldn't be able to change the condition type to Chef Policy Group.

Comment (tylercloke) [reply components/authz-service/server/v2/rules_test.go]: Sure!

File: components/authz-service/server/v2/rules_test.go [hunk 4]
```diff
@@ -157,6 +157,177 @@ func TestCreateRule(t *testing.T) {
...
+				Id:         id,
+				Name:       "my other foo",
+				ProjectId:  "foo-project",
+				Conditions: changedAPIConditions,
+				Type:       api.ProjectRuleTypes_EVENT,
+			})
+			grpctest.AssertCode(t, codes.InvalidArgument, err)
+			assert.Nil(t, resp)
+		}},
+		{"with valid rule data, returns no error and updates the rule in storage", func(t *testing.T) {
```
Comment (msorens) [inline components/authz-service/server/v2/rules_test.go]: This test is for a Node rule type with a condition **not** available for Event type.
A couple more useful tests:
Same test for **Node** rule type with a condition that **is** available for Event type.
Same test for **Event** rule type with a condition that is available for Node type.
(We cannot do the 4th possibility:  test for Event with condition not available for Node.)

(Also, if you have time, would be useful to have similar tests for CreateRule.)

**Pull request merged**